File: swift/problem155.swift
//ref: https://www.geeksforgeeks.org/design-a-stack-that-supports-getmin-in-o1-time-and-o1-extra-space/
class MinStack {
    private var stack: [Int]
    private var minElement: Int
    
    init() {
        self.stack = [Int]()
        self.minElement = Int.min
    }
    
    func push(_ x: Int) {
        if stack.isEmpty {
            stack.append(x)
            minElement = x
        } else if (x >= minElement) {
            stack.append(x)
        } else {
            stack.append(2*x-minElement)
            minElement = x
        }
    }
    
    func pop() {
        let tmp = stack.removeLast()
        if (tmp < minElement) {
            minElement = 2*minElement - tmp
        }
    }
    
    func top() -> Int {
        let tmp = stack.last!
        return tmp >= minElement ? tmp : 2*minElement - tmp
    }
    
    func getMin() -> Int {
        return minElement
    }
}

File: python/problem155.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.li = []
        

    def push(self, x: int) -> None:
        self.li.append(x)
        

    def pop(self) -> None:
        self.li.pop()
        

    def top(self) -> int:
        return self.li[-1]
        

    def getMin(self) -> int:
        return min(self.li)
        
        
        
# https://leetcode.com/problems/min-stack/discuss/236242/Python3-52-ms-100-Deque
class MinStack:

    def __init__(self):
        self.dq = collections.deque([None])

    def push(self, x: 'int') -> 'None':
        if self.dq[0] == None or self.dq[0] >= x:
            self.dq.appendleft(x)
        self.dq.append(x)

    def pop(self) -> 'None':
        if self.dq[0] != None and self.dq[-1] <= self.dq[0]:
            self.dq.popleft()
        return self.dq.pop()

    def top(self) -> 'int':
        return self.dq[-1]

    def getMin(self) -> 'int':
        return self.dq[0]